packages/core/database/lib/entity-manager/relations-orderer.js

```diff
@@ -1,6 +1,106 @@
+ *   { id: 5, position: { before: 1 } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 2, position: { end: true } }
+ *
+ * Going through the connect array, id 5 has to be connected before id 1,
+ * so the order of id5 = id1 - 1. But the order value of id 1 is unknown.
+ * The only way to know the order of id 1 is to connect it first.
+ *
+ * This function makes sure the relations are connected in the right order:
+ *   { id: 2, position: { end: true } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 5, position: { before: 1 } }
+ *
+ */
+const sortConnectArray = (connectArr, initialArr = [], strictSort = true) => {
+  const sortedConnect = [];
+  // Boolean to know if we have to recalculate the order of the relations
+  let needsSorting = false;
+  // Map to validate if relation is already in sortedConnect or DB.
+  const relInArray = initialArr.reduce((acc, rel) => ({ ...acc, [rel.id]: true }), {});
+  // Map to store the first index where a relation id is connected
+  const firstSeen = {};
+  // Map to validate if connect relation has already been computed
+  const computedIdx = {};
+
+  connectArr.forEach((rel, idx) => {
+    // If adjacent relation is not in the database or seen yet in the connect array
+    // then we need to sort the connect array
+    const adjacentRelId = rel.position?.before || rel.position?.after;
+    if (!relInArray[adjacentRelId] && !firstSeen[adjacentRelId]) needsSorting = true;
+    // Populate firstSeen
+    if (!(rel.id in firstSeen)) firstSeen[rel.id] = idx;
+  });
+
+  // If we don't need to sort the connect array, we can return it as is
+  if (!needsSorting) return connectArr;
+
+  // Add relation to sortedConnect and mark it as seen
+  const pushRelation = (rel) => {
+    sortedConnect.push(rel);
+    relInArray[rel.id] = true;
+  };
+
+  // Recursively compute in which order the relation should be connected
+  const computeRelation = (rel, idx, relationsSeenInBranch) => {
+    const adjacentRelId = rel.position?.before || rel.position?.after;
+
+    // This connect has already been computed
+    if (idx in computedIdx) return;
+
+    if (!adjacentRelId || relInArray[adjacentRelId]) {
+      return pushRelation(rel);
+    }
+
+    // If the relation has already been seen in the current branch,
+    // it means there is a circular reference
+    if (relationsSeenInBranch[adjacentRelId]) {
+      throw new InvalidRelationError(
+        'A circular reference was found in the connect array. ' +
+          'One relation is trying to connect before/after another one that is trying to connect before/after it'
+      );
+    }
+
+    // Look if id is referenced elsewhere in the array
+    const adjacentRelIdx = firstSeen[adjacentRelId];
+
+    if (adjacentRelIdx) {
+      const adjacentRel = connectArr[adjacentRelIdx];
+
+      // Mark adjacent relation idx as computed,
+      // so it is not computed again later in the loop
+      computedIdx[adjacentRelIdx] = true;
```

Comment:
Can we use the same mapper as firstSeen to store this information? 🤔 
Would be like:
```javascript
{
  1: { index: 23, computed: false },
}
```
And then set `adjacentRelIdx.computed = true;` ?

Reply:
💡  sounds good

Reply:
actually , we may not be able to merge it. As today, relations id can be repeated in the same list, so you could have two indexes that have not been computed.

Reply:
But didn't you already override the duplicated indexes with this code? 🤔

Reply:
Nope :/ , duplicated indexes are allowed right now.